File: SCALA/src/cars/FFcar.scala
/*
 * FFcar.scala
 * package cars
 * Author: Luca Ghislotti
 *
 * Classe FFcar ottenuta come sottoclasse dalla superclasse
 * Car, definita come interfaccia (Scala trait)
 *
 */

package cars

/*
 * Sottoclasse costruita sfruttando il costruttore della superclasse Car,
 * a cui vengono aggiunti i campi specifici per le auto tradizionali
 */
class FFcar(targa: String, potenza: Int, peso: Int, var capacitaSerbatoio: Int = 50, urbano: Double = 4.5,
  combi: Double = 5.6, extra: Double = 7.3) extends Car(targa, potenza, peso) {
    
  /*
   * Creazione mappa per effettuare l'associazione tra
   * la denonminazione del consumo ed il valore vero e proprio
   */
  val consumi = Map(
    "   Urbano" -> urbano,
    "Combinato" -> combi,
    "    Extra" -> extra)

  /*
   * Metodi getter e setter per il campo
   * capacitaSerbatoio
   */
  def getCapSerb: Int = capacitaSerbatoio // getter
  def setCapSerb(newCapSer: Int) = capacitaSerbatoio = newCapSer // setter

  /*
   * Funzione per effettuare la stampa dell'auto FFcar
   *
   * Da notare l'utilizzo di funzioni innestate: si costruisce
   * internamente una funzione per la stampa dei consumi
   *
   * Da notare la clausola "override" che permette di ridefinire
   * il medesimo metodo della superclasse
   */
  override def showInfo {
    /*
     * Funzione innestata per effettuare la stampa dei consumi
     * La mappa viene iterata tramite ciclo foreach restituendo
     * la coppia <denominazione, consumo>
     */
    def printResumee {
      consumi.foreach {
        case (denominazione, valore) => println(s"$denominazione:  $valore L/100km")
      }
    }

    print("FOSSILE FUEL ")
    super.showInfo() // utilizzo del costrutto super per richiamare il metodo della superclasse
    println("Cap. Ser.:  " + getCapSerb + " L")
    printResumee
    println("")
  }
}

File: QT/Info3_progettoCPP/FFcar.cpp
/*
 * FFcar.cpp
 * Author: Luca Ghislotti
 *
 * Implementazione della classe FFcar
 *
 */

#include <iostream>
using namespace std;
#include <string>
#include "FFcar.h"

FFcar::FFcar(char *targaInput, int pesoInput, int potenzaInput,
        short fuelCapacity, double urbano, double combinato, double extra) :
        Car(targaInput, pesoInput, potenzaInput), capacitaSerbatoio(
                fuelCapacity) {
    consumi.insert(make_pair("    Urbano ", urbano));
    consumi.insert(make_pair("    Combi. ", combinato));
    consumi.insert(make_pair("      Extra", extra));
}

short FFcar::getCapSerb() {
    return capacitaSerbatoio;
}

string FFcar::printResumee() {
    string resumee = "";
    for (auto it = consumi.cbegin(); it!=consumi.cend(); it++)
        resumee = resumee + it->first + ":  " + std::to_string(it->second) +" L/100km\n";

    return resumee;
}

string FFcar::showInfo() {
    return "\nFOSSILE FUEL " + Car::showInfo() + "  Cap. Ser.:  " + std::to_string(getCapSerb()) + " L" + "\n" + printResumee();
}

string FFcar::getClassName() {
    return "F";
}

FFcar::~FFcar() {
    consumi.clear();
    cout << "cleared all FFcars" << endl;
}
